### PR TITLE
[gdal] Add open options support to gain dataset flexibility / fixes

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds open options support to the gdal provider. While it might look like a lot of code, most of it is actually copy / pasted from @rouault 's implementation for the ogr provider which has been in QGIS since 3.16.

The primary motivation here is to fix a situation with GeoPDFs whereas those would open at very bad resolution within QGIS even though the GeoPDF itself had been exported using a DPI of 300. There's been related issues filed on this both in QGIS (https://github.com/qgis/QGIS/issues/33465) and GDAL (https://github.com/OSGeo/gdal/pull/2961). 

Here's a bad resolution vs. fixed resolution comparison for GeoPDFs:

https://user-images.githubusercontent.com/1728657/106376211-a56ca500-63c5-11eb-834f-a1a9ce1f257b.mp4

 Yep, that bad.

UI wise, this PR re-implements what was in the OGR source select. It looks like this:
image

Acknowledging that while I see this as a significant fix, it definitively encroaches into feature territory. As such, I would like to request a freeze exemption on the basis of the code fixing a bad situation as well as it having bee tested via the ogr provider for a while now.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Code looks good to me. I'm +1 for merging this as a bug fix, given that we did the exact same thing when the vector counterpart was added (under the rationale that many bugs could be attribute to a lack of ability to change these gdal options). I'll let @m-kuhn make the final call here though!

File: src/gui/providers/gdal/qgsgdalsourceselect.cpp [hunk 1]
```diff
@@ -189,13 +234,157 @@ void QgsGdalSourceSelect::addButtonClicked()
...
+
+void QgsGdalSourceSelect::fillOpenOptions()
+{
+  clearOpenOptions();
+  computeDataSources();
+  if ( mDataSources.isEmpty() )
+    return;
+
+  GDALDriverH hDriver;
+  if ( STARTS_WITH_CI( mDataSources[0].toUtf8().toStdString().c_str(), "PG:" ) )
```
Comment (rouault) [inline src/gui/providers/gdal/qgsgdalsourceselect.cpp:271]: This check should be removed. For the raster side, it is expected that the PostgisRaster driver triggers for a GDAL dataset name starting with PG:

File: src/gui/providers/gdal/qgsgdalsourceselect.cpp [hunk 2]
```diff
@@ -189,13 +234,157 @@ void QgsGdalSourceSelect::addButtonClicked()
...
+    if ( pszOptionName == nullptr )
+      continue;
+
+    // Exclude options that are not of raster scope
+    const char *pszScope = CPLGetXMLValue( psItem, "scope", nullptr );
+    if ( pszScope != nullptr && strstr( pszScope, "raster" ) == nullptr )
+      continue;
+
+    // The GPKG driver list a lot of options that are only for rasters
+    if ( bIsGPKG && strstr( pszOpenOptionList, "scope=" ) == nullptr &&
```
Comment (rouault) [inline src/gui/providers/gdal/qgsgdalsourceselect.cpp:309]: that particular case valid for vector is not for raster

**Comment by nirvn** [comment]
@rouault , comments addressed. Big thanks for providing most of the code through what you added for the OGR provider one cycle ago.

**Review by m-kuhn — APPROVED** [review]
Looks like something we want to go through during feature freeze

**Pull request merged**